src/test/java/io/confluent/examples/streams/CustomStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,538 @@
+                context.forward(record.key, joinedValue);
+                streamBufferStore.delete(record.key);
+              }
+            }
+          }
+        }
+
+        private boolean waitTimeExpired(final Instant recordTimestamp, final long currentStreamTime) {
+          return Duration.between(recordTimestamp, Instant.ofEpochMilli(currentStreamTime))
+            .compareTo(approxMaxWaitTimePerRecordForTableData) > 0;
+        }
+
+        @Override
+        public void close() {
+        }
+
+      };
+    }
+
+  }
+
+  /**
+   * Implements table-side triggering of join output.  This is but one example for such custom join semantics -- feel
+   * free to modify this example to match your own needs.
+   *
+   * For every <i>observed</i> record arriving at its upstream table, this transformer will check for a buffered (i.e.,
+   * not yet joined) record on the stream side.  If there is a match, then the transformer will produce a fully
+   * populated join output message -- which is the desired table-side triggering behavior.  If there is no match, then
+   * the transformer will effectively do nothing (it will actually send a tombstone record).
+   */
+  private static final class StreamTableJoinTableSideLogic
+    implements ValueTransformerWithKeySupplier<String, Long, Pair<Double, Long>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamTableJoinTableSideLogic.class);
+
+    private final Duration approxMaxWaitTimePerRecordForTableData;
+    private final String streamBufferStoreName;
+
+    StreamTableJoinTableSideLogic(final Duration approxMaxWaitTimePerRecordForTableData,
+                                  final String streamBufferStoreName) {
+      this.approxMaxWaitTimePerRecordForTableData = approxMaxWaitTimePerRecordForTableData;
+      this.streamBufferStoreName = streamBufferStoreName;
+    }
+
+    @Override
+    public ValueTransformerWithKey<String, Long, Pair<Double, Long>> get() {
+      return new ValueTransformerWithKey<String, Long, Pair<Double, Long>>() {
+
+        private KeyValueStore<String, Pair<Double, Instant>> streamBufferStore;
+        private ProcessorContext context;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+          this.context = context;
+        }
+
+        /**
+         * A return value of `null` for this method represents a table tombstone record.  We need to discard such
+         * tombstone records in the downstream processing topology (see code further above in this file) because, in the
+         * context of this example of a custom stream-table join, they must not appear in the join output.  Note that we
+         * can't discard the tombstones directly in the ValueTransformerWithKey.
```

Comment:
I added a note about `transformValues()` vs. `transform()`, but kept `transformValues()` to avoid the potential re-partitioning step.

Also, I added the TODO.  I suppose you meant `flatTransformValues()` though. ;-)